### ISSUE TITLE
Convex Auth consumes pages that have "code" as search param

### ISSUE THREAD

**Issue by tomredman** [get-convex/convex-auth#145]
Came across this curious case recently:
- A user is created and authenticated through Convex Auth, no problem, no sweat.
- As a business rule, that user needs to be able to connect a Mailchimp account to their profile (entirely independent of Convex Auth)
- Kick off Mailchimp's oauth flow and upon success, it redirects back to my app, where we're using Convex Auth (e.g. https://my-sick-app.com/app/integrations/mailchimp/connect?code=12345)

<img width="478" alt="image" src="https://github.com/user-attachments/assets/5ea88618-a23a-4aa2-9b8e-d81261e50751" />

After some head scratching, I realized my ConvexAuthProvider was snatching up the `?code=12345` portion of my URL, attempting to authenticate it as if it were a sanctioned auth provider, failing, clearing my locally stored auth tokens, ultimately leaving me in a logged-out state.

Unfortunately this all happens prior to my /integrations/mailchimp/connect page loading, so I can't "block" it there, or do anything in that place. 

I'm sure I can find a workaround, but it'd be great to somehow (not sure how!) tell Convex Auth not to worry about this `?code`, I got it.

<img width="845" alt="image" src="https://github.com/user-attachments/assets/558964a6-6837-4012-8b9c-1c80b23d21c3" />

**Comment by thomasballinger** [comment]
Sounds like we should be using state= param to the auth request to GitHub, etc

**Comment by thomasballinger** [comment]
Today the Convex Auth middleware sees the `?code=` query parameter and assumes the request is intended to be a new login.

I wonder if we can use a more specific query parameter than `?code=` to distinguish this, say `convexAuthCode=`. The OAuth provider's specified redirect (which must contain a `?code=` query parameter) sends the browser to the Convex backend, so skips this middleware completely. The code query parameter for the next redirect this is specific to this library.

I need to confirm this, there are other flows but they might all be able to use a different name.

If this doesn't work, we could register certain routes and not oauth-redirectable-to and those routes don't slurp up ?code=
or register certain routes as oauth-redirectable-to and only those routes can be redirected to, checking at runtime so there's a clear error if you do.

**Comment by tomredman** [comment]
Oh, that's good news then @thomasballinger. For now, I have a simple hack that is working well by rewriting the search param in `berforeLoad`, so there is no urgency on my end:

```
export const Route = createFileRoute("/_authenticated/integrations/mailchimp")({
  loaderDeps: ({ search: { code } }) => ({ code }),
  beforeLoad: ({ search }) => {
    console.log(search);
    if (search.code) {
      throw redirect({
        to: "/integrations/mailchimp",
        search: {
          mailchimpCode: search.code,
          code: undefined,
        },
      });
    }
  },
  component: MailchimpComponent,
  validateSearch: (search: Record<string, unknown>): MailchimpSearch => {
    return {
      code: (search.code as string) || undefined,
      mailchimpCode: (search.mailchimpCode as string) || undefined,
    };
  },
});
```

**Comment by thomasballinger** [comment]
After some investigation here, the name of the `code=` query parameter is out of our hands for things like Resend verification emails, which we want to be able to send you back your site URL instead of your Convex backend site URL so that SPA auth flows work well.

The most pragmatic thing seems to be a list of URL patterns to ignore or a callback to examine the request in middleware to know to skip it. This would be a change to middleware, a callback or pattern you pass in that returns true or matches when the `?code=` query parameter should be ignored, but normal auth logic should still apply (the Next.js route should still be protected, an user not signed in will still be e.g. redirected to /sign-in).

Any design thoughts welcome, I'm going to hit some more urgent auth things first but if this blocking anyone else please let me know and we can reprioritize.

**Comment by sshader** [comment]
> the name of the code= query parameter is out of our hands for things like Resend verification emails

I don't think this is true -- we're just hardcoding that the parameter is named `code` without allowing any customization [here](https://github.com/get-convex/convex-auth/blob/66e1197a295ea5365bd7b64469e2be73d8fdc208/src/server/implementation/signIn.ts#L146)

So an alternative option here would be allowing the name of the query param to be customizable (but still intercepting the query param on every request). But I think it might be more straightforward to allow configuring which routes should expect a `code` query param and which to leave alone.

**Comment by paintlikeanengineer** [comment]
Hey Tom, this is now blocking me. I don't use TanStack Router and so don't have the beforeLoad prop that I can use to send a replace query param function.

**Comment by munj1** [comment]
> Hey Tom, this is now blocking me. I don't use TanStack Router and so don't have the beforeLoad prop that I can use to send a replace query param function.
> 
> I do this in middleware.ts but am unable to intercept.
> 
> ```
> export default convexAuthNextjsMiddleware((request) => {
>   const url = new URL(request.url)
>   console.log('Middleware full URL:', url.href)
> 
>   // Intercept the Figma callback route to rewrite "code" as "figmaCode"
>   if (url.pathname.startsWith('/settings/profile/figma-auth-callback') && url.searchParams.has('code')) {
>     const code = url.searchParams.get('code')!
>     url.searchParams.delete('code')
>     url.searchParams.set('figmaCode', code)
>     // Redirect to the same URL with the renamed query parameter.
>     return NextResponse.redirect(url)
>   }
>   if (!isPublicPage(request) && !isAuthenticatedNextjs()) {
>     return nextjsMiddlewareRedirect(request, '/auth')
> })
> ```

Same issue here
Can't intercepts code params under convexAuthNextjsMiddleware

**Comment by Maris1984** [comment]
Is there perhaps a fix for ones using React and not using the TanStack Router?

**Comment by vhgn** [comment]
can confirm, this is fixed on the latest version (0.0.90)
